In [3]:
# Price Data

item_dict = {'almonds': 6,
 'antioxidant juice': 9,
 'avocado': 1,
 'black tea': 4,
 'burgers': 4,
 'champagne': 3,
 'chicken': 4,
 'chocolate': 7,
 'chutney': 7,
 'cookies': 7,
 'cooking oil': 7,
 'cottage cheese': 9,
 'eggs': 9,
 'energy bar': 9,
 'energy drink': 3,
 'extra dark chocolate': 8,
 'french fries': 7,
 'fresh tuna': 9,
 'frozen smoothie': 7,
 'frozen vegetables': 2,
 'green grapes': 8,
 'green tea': 8,
 'honey': 9,
 'light cream': 5,
 'low fat yogurt': 4,
 'meatballs': 9,
 'milk': 5,
 'mineral water': 2,
 'oil': 9,
 'olive oil': 8,
 'pet food': 3,
 'protein bar': 1,
 'salad': 7,
 'salmon': 3,
 'shallot': 6,
 'shrimp': 2,
 'soup': 2,
 'spaghetti': 6,
 'spinach': 7,
 'tomato juice': 8,
 'tomatoes': 5,
 'turkey': 8,
 'vegetables mix': 3,
 'whole wheat flour': 5,
 'whole wheat pasta': 5,
 'whole wheat rice': 4,
 'yams': 7}

In [4]:
# transaction data

transaction_data = \
[['shrimp', 'almonds', 'avocado', 'vegetables mix', 'green grapes', 'whole wheat flour', 'yams', 'cottage cheese', 'energy drink', 'tomato juice', 'low fat yogurt', 'green tea', 'honey', 'salad', 'mineral water', 'salmon', 'antioxidant juice', 'frozen smoothie', 'spinach', 'olive oil'],
['burgers', 'meatballs', 'eggs'],
['chutney'],
['turkey', 'avocado'],
['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
['low fat yogurt'],
['whole wheat pasta', 'french fries'],
['soup', 'light cream', 'shallot'],
['frozen vegetables', 'spaghetti', 'green tea'],
['french fries'],
['eggs', 'pet food'],
['cookies'],
['turkey', 'burgers', 'mineral water', 'eggs', 'cooking oil'],
['spaghetti', 'champagne', 'cookies'],
['mineral water', 'salmon'],
['mineral water'],
['shrimp', 'chocolate', 'chicken', 'honey', 'oil', 'cooking oil', 'low fat yogurt'],
['turkey', 'eggs'],
['turkey', 'fresh tuna', 'tomatoes', 'spaghetti', 'mineral water', 'black tea', 'salmon', 'eggs', 'chicken', 'extra dark chocolate'],
['meatballs', 'milk', 'honey', 'french fries', 'protein bar']]

1. Find how many unique items are present in the data?  (Hint: It will correspond to the number of key pairs in the Cost dictionary)

In [5]:
unique_items = len(item_dict.keys())
unique_items

47

2. Find the total number of transactions?

In [6]:
total_transactions = len(transaction_data)
total_transactions

20

3. Add items, try adding the item “banana” in stores. The cost of this item is '$5'. Add the item “banana” to the Cost dictionary.

In [7]:
item_dict["banana"] = 5


4. Use a function named add_item( ) to automatically add an item to the Cost dictionary given its name and price.

In [8]:
def add_item(item,price):
    '''
    Adds new item with the price to the item_dict
    
    Parameters: 
    item (string) : Name of the item
    price (int) : Price of the item
    
    Returns:
    Dictionary : Returns item_dict
    '''
    item_dict[item]=price
    return item_dict

5. Create a function called remove_item( ) to automatically remove an item from the Cost dictionary given its name.

In [9]:
def remove_item(item):
    '''
    Removes the given item from the item_dict
    
    Parameters: 
    item (string) : Name of the item
    
    Returns:
    Dictionary : Returns item_dict
    '''
    
    item_dict.pop(item)
    return item_dict

Question 2

1. Write the code, to find the price of an item from the cost dictionary

    Find the price of salmon from the cost dictionary

In [10]:
price_salmon = item_dict["salmon"]
price_salmon

3

2. Write a code that helps the administrator find any particular transaction in the sequence

    Try finding the ninth transaction from the data

In [11]:
transaction_9 = transaction_data[8]
transaction_9

['frozen vegetables', 'spaghetti', 'green tea']

3. How many items are present in the ninth transaction?

In [12]:
no_items_transaction_9 = len(transaction_9)
no_items_transaction_9

3

4. Find the total cost of the ninth transaction from the array of transactions.

In [13]:
total_cost_9_transaction = 0

for item in transaction_9:
    total_cost_9_transaction+=item_dict[item]

total_cost_9_transaction

16

5. Create a function named find_total( ) to find the final price of any transaction after applicable taxes have been implemented.

In [14]:
def find_total(transaction):
    '''
    Finds total cost of particular transaction from transaction data
    
    Parameters: 
    transaction (List) : List of items of a particular transaction
    
    Returns:
    float : Returns total cost of the particular transaction
    '''
    total_cost = 0
    for item in transaction:
        total_cost+=item_dict[item]
    tax = 0.05*total_cost
    total_cost +=tax
    return total_cost


6. As per government guidelines, a mandatory 5% tax is applied to the total cost of a transaction. Find the final price of the second transaction after applicable taxes have been implemented.

In [15]:
find_total(transaction_data[1])

23.1

Question 3.

1. What is the data structure that can be used to count the frequency of items purchased (from the transaction data)?
   Ans: Dictionary

Ans: Dictionary

2. Create a frequency counter to count the number of items:

In [16]:
def calculate_frequency():
    '''
    Will count the total number of each items purchased from the transaction data
    
    Returns:
    Dictionary : Returns frequency dictionary
    '''
    frequency_dict = {}  

    for transaction in transaction_data:
        for item in transaction:
            if item in frequency_dict:
                frequency_dict[item]+=1
            else:
                frequency_dict[item]=1
    return frequency_dict
frequency_dict = calculate_frequency()
frequency_dict

{'shrimp': 2,
 'almonds': 1,
 'avocado': 2,
 'vegetables mix': 1,
 'green grapes': 1,
 'whole wheat flour': 1,
 'yams': 1,
 'cottage cheese': 1,
 'energy drink': 1,
 'tomato juice': 1,
 'low fat yogurt': 3,
 'green tea': 3,
 'honey': 3,
 'salad': 1,
 'mineral water': 6,
 'salmon': 3,
 'antioxidant juice': 1,
 'frozen smoothie': 1,
 'spinach': 1,
 'olive oil': 1,
 'burgers': 2,
 'meatballs': 2,
 'eggs': 5,
 'chutney': 1,
 'turkey': 4,
 'milk': 2,
 'energy bar': 1,
 'whole wheat rice': 1,
 'whole wheat pasta': 1,
 'french fries': 3,
 'soup': 1,
 'light cream': 1,
 'shallot': 1,
 'frozen vegetables': 1,
 'spaghetti': 3,
 'pet food': 1,
 'cookies': 2,
 'cooking oil': 2,
 'champagne': 1,
 'chocolate': 1,
 'chicken': 2,
 'oil': 1,
 'fresh tuna': 1,
 'tomatoes': 1,
 'black tea': 1,
 'extra dark chocolate': 1,
 'protein bar': 1}

In [17]:
def update_counter(new_transaction):
    '''
    Will update the frequency dictionary for the new tansaction
    
    Parameters:
    new_transaction (List) : List of items in new transaction
    
    Returns:
    Dictionary : Returns updated frequency dictionary
    '''
    for item in new_transaction:
        if item in frequency_dict:
            frequency_dict[item]+=1
        else:
            frequency_dict[item]=1
    return frequency_dict


3. A new transaction has just been made. This transaction includes “honey,” “pet food,” and “shrimp.” Update the frequency counter to reflect this new transaction.

In [18]:
frequency_counter = update_counter([ 'honey', 'pet food','shrimp'])
frequency_counter

{'shrimp': 3,
 'almonds': 1,
 'avocado': 2,
 'vegetables mix': 1,
 'green grapes': 1,
 'whole wheat flour': 1,
 'yams': 1,
 'cottage cheese': 1,
 'energy drink': 1,
 'tomato juice': 1,
 'low fat yogurt': 3,
 'green tea': 3,
 'honey': 4,
 'salad': 1,
 'mineral water': 6,
 'salmon': 3,
 'antioxidant juice': 1,
 'frozen smoothie': 1,
 'spinach': 1,
 'olive oil': 1,
 'burgers': 2,
 'meatballs': 2,
 'eggs': 5,
 'chutney': 1,
 'turkey': 4,
 'milk': 2,
 'energy bar': 1,
 'whole wheat rice': 1,
 'whole wheat pasta': 1,
 'french fries': 3,
 'soup': 1,
 'light cream': 1,
 'shallot': 1,
 'frozen vegetables': 1,
 'spaghetti': 3,
 'pet food': 2,
 'cookies': 2,
 'cooking oil': 2,
 'champagne': 1,
 'chocolate': 1,
 'chicken': 2,
 'oil': 1,
 'fresh tuna': 1,
 'tomatoes': 1,
 'black tea': 1,
 'extra dark chocolate': 1,
 'protein bar': 1}

Question 4.

1. "What is the top-selling item?"

In [19]:
sort_by_values = sorted(frequency_counter.items(),key=lambda x:(x[1],x[0]),reverse=True)
top_3_selling_items = dict(sort_by_values[:3])
top_3_selling_items

{'mineral water': 6, 'eggs': 5, 'turkey': 4}

In [20]:
top_selling_item = sort_by_values[0][0]
top_selling_item

'mineral water'

In [21]:
print(dict(sort_by_values))

{'mineral water': 6, 'eggs': 5, 'turkey': 4, 'honey': 4, 'spaghetti': 3, 'shrimp': 3, 'salmon': 3, 'low fat yogurt': 3, 'green tea': 3, 'french fries': 3, 'pet food': 2, 'milk': 2, 'meatballs': 2, 'cooking oil': 2, 'cookies': 2, 'chicken': 2, 'burgers': 2, 'avocado': 2, 'yams': 1, 'whole wheat rice': 1, 'whole wheat pasta': 1, 'whole wheat flour': 1, 'vegetables mix': 1, 'tomatoes': 1, 'tomato juice': 1, 'spinach': 1, 'soup': 1, 'shallot': 1, 'salad': 1, 'protein bar': 1, 'olive oil': 1, 'oil': 1, 'light cream': 1, 'green grapes': 1, 'frozen vegetables': 1, 'frozen smoothie': 1, 'fresh tuna': 1, 'extra dark chocolate': 1, 'energy drink': 1, 'energy bar': 1, 'cottage cheese': 1, 'chutney': 1, 'chocolate': 1, 'champagne': 1, 'black tea': 1, 'antioxidant juice': 1, 'almonds': 1}


Insertion Sort

In [22]:
import time
start = time.time()
def insertion_sort(frequency_counter):
    frequency_list = list(frequency_counter.items())  
    l = len(frequency_list)
    for i in range(1,l):
        key = frequency_list[i]
        
        j=i-1
        
        while j>=0 and frequency_list[j][1]<key[1]:
            
            frequency_list[j+1]=frequency_list[j]
            j-=1
            
        frequency_list[j+1]=key
    return dict(frequency_list)
        
print(insertion_sort(frequency_counter))
end = time.time()
total_time_insertion = (end-start) *1000

{'mineral water': 6, 'eggs': 5, 'honey': 4, 'turkey': 4, 'shrimp': 3, 'low fat yogurt': 3, 'green tea': 3, 'salmon': 3, 'french fries': 3, 'spaghetti': 3, 'avocado': 2, 'burgers': 2, 'meatballs': 2, 'milk': 2, 'pet food': 2, 'cookies': 2, 'cooking oil': 2, 'chicken': 2, 'almonds': 1, 'vegetables mix': 1, 'green grapes': 1, 'whole wheat flour': 1, 'yams': 1, 'cottage cheese': 1, 'energy drink': 1, 'tomato juice': 1, 'salad': 1, 'antioxidant juice': 1, 'frozen smoothie': 1, 'spinach': 1, 'olive oil': 1, 'chutney': 1, 'energy bar': 1, 'whole wheat rice': 1, 'whole wheat pasta': 1, 'soup': 1, 'light cream': 1, 'shallot': 1, 'frozen vegetables': 1, 'champagne': 1, 'chocolate': 1, 'oil': 1, 'fresh tuna': 1, 'tomatoes': 1, 'black tea': 1, 'extra dark chocolate': 1, 'protein bar': 1}


In [23]:
print("Total execution time in milliseconds for insertion sort = ", total_time_insertion)

Total execution time in milliseconds for insertion sort =  0.9980201721191406


Selection Sort

In [24]:
import time
start = time.time()
def selection_sort(frequency_counter):
    frequency_list = list(frequency_counter.items())  
    l = len(frequency_list)
    for i in range(l):
        maxi = i
        for j in range(i+1,l):
            if frequency_list[j][1]>frequency_list[maxi][1]:
                maxi=j
        frequency_list[i],frequency_list[maxi] = frequency_list[maxi],frequency_list[i]
    sortdict=dict(frequency_list)
    return sortdict
print(selection_sort(frequency_counter))
end = time.time()
total_time_selection = (end-start)*1000

{'mineral water': 6, 'eggs': 5, 'honey': 4, 'turkey': 4, 'low fat yogurt': 3, 'green tea': 3, 'shrimp': 3, 'salmon': 3, 'french fries': 3, 'spaghetti': 3, 'avocado': 2, 'burgers': 2, 'meatballs': 2, 'milk': 2, 'pet food': 2, 'cookies': 2, 'cooking oil': 2, 'chicken': 2, 'spinach': 1, 'olive oil': 1, 'whole wheat flour': 1, 'green grapes': 1, 'almonds': 1, 'chutney': 1, 'vegetables mix': 1, 'salad': 1, 'energy bar': 1, 'whole wheat rice': 1, 'whole wheat pasta': 1, 'energy drink': 1, 'soup': 1, 'light cream': 1, 'shallot': 1, 'frozen vegetables': 1, 'tomato juice': 1, 'yams': 1, 'cottage cheese': 1, 'antioxidant juice': 1, 'champagne': 1, 'chocolate': 1, 'frozen smoothie': 1, 'oil': 1, 'fresh tuna': 1, 'tomatoes': 1, 'black tea': 1, 'extra dark chocolate': 1, 'protein bar': 1}


In [25]:
print("Total execution time in milliseconds for selection sort = ", total_time_selection)

Total execution time in milliseconds for selection sort =  0.9965896606445312


Merge Sort

In [26]:
import time
start = time.time()
frequency_list = list(frequency_counter.items())
def merge_sort(frequency_list):
    if len(frequency_list)>1:
        l = len(frequency_list)
        mid = l//2
        L=frequency_list[:mid]
        R=frequency_list[mid:]
        merge_sort(L)
        merge_sort(R)
        i=j=k=0
        while i<len(L) and j<len(R):
            if L[i][1]>R[j][1]:
                frequency_list[k]=L[i]
                i+=1
            else:
                frequency_list[k]=R[j]
                j+=1
            k+=1
        while i<len(L):
            frequency_list[k]=L[i]
            i+=1
            k+=1
        while j<len(R):
            frequency_list[k]=R[j]
            j+=1
            k+=1
        return dict(frequency_list)
print(merge_sort(frequency_list))
end = time.time()
total_time_merge = (end-start)*1000

{'mineral water': 6, 'eggs': 5, 'turkey': 4, 'honey': 4, 'spaghetti': 3, 'french fries': 3, 'salmon': 3, 'green tea': 3, 'low fat yogurt': 3, 'shrimp': 3, 'chicken': 2, 'cooking oil': 2, 'cookies': 2, 'pet food': 2, 'milk': 2, 'meatballs': 2, 'burgers': 2, 'avocado': 2, 'protein bar': 1, 'extra dark chocolate': 1, 'black tea': 1, 'tomatoes': 1, 'fresh tuna': 1, 'oil': 1, 'chocolate': 1, 'champagne': 1, 'frozen vegetables': 1, 'shallot': 1, 'light cream': 1, 'soup': 1, 'whole wheat pasta': 1, 'whole wheat rice': 1, 'energy bar': 1, 'chutney': 1, 'olive oil': 1, 'spinach': 1, 'frozen smoothie': 1, 'antioxidant juice': 1, 'salad': 1, 'tomato juice': 1, 'energy drink': 1, 'cottage cheese': 1, 'yams': 1, 'whole wheat flour': 1, 'green grapes': 1, 'vegetables mix': 1, 'almonds': 1}


In [27]:
print("Total execution time in milliseconds for merge sort = ", total_time_merge)

Total execution time in milliseconds for merge sort =  0.9887218475341797


Ranking of sorting algorithm based on Execution Time(Highest to Lowest)
    1. Insertion Sort - 0.9980201721191406 ms
    2. Selection Sort - 0.9965896606445312 ms
    3. Merge Sort - 0.9887218475341797 ms
    